File: ShellLib/Size.cpp
#include "Stdafx.h"
#include "Size.h"

namespace ShellLib { namespace Drawing {

#pragma	region	Constructor
	/*********************
	*	Constructor
	**********************/

	CSize::CSize(CCoord c) {
		m_width = c.X;
		m_height = c.Y;
	}

	CSize::CSize(short w, short h) {
		m_width = w;
		m_height = h;
	}

#pragma endregion


#pragma region Properties
	/******************
	* Properties
	*******************/

	bool	CSize::IsEmpty::get(void) {
		return m_width == m_height == 0;
	}

	short	CSize::Height::get(void) {
		return m_height;
	}
	void	CSize::Height::set(short value) {
		m_height = value;
	}

	short	CSize::Width::get(void) {
		return m_width;
	}
	void	CSize::Width::set(short value) {
		m_width = value;
	}

#pragma endregion


#pragma	region	Operations
	/*******************
	*    Operations    *
	*******************/

	CSize::operator ShellLib::Drawing::CCoord(CSize sz) {
		return CCoord(sz.m_width, sz.m_height);
	}

	CSize	CSize::operator+(CSize left, CSize right) {
		return Add(left, right);
	}

	CSize	CSize::operator-(CSize left, CSize right) {
		return Subtract(left, right);
	}

	bool	CSize::operator==(CSize left, CSize right) {
		return left.m_width == right.m_width && left.m_height == right.m_height;
	}

	bool	CSize::operator!=(CSize left, CSize right) {
		return !(left == right);
	}

	CSize	CSize::Add(CSize sz1, CSize sz2) {
		return CSize(sz1.m_width + sz2.m_width, sz1.m_height + sz2.m_height);
	}

	CSize	CSize::Subtract(CSize sz1, CSize sz2) {
		return CSize(sz1.m_width - sz2.m_width, sz1.m_height - sz2.m_height);
	}

	bool	CSize::Equals(Object ^ obj) {
		if (obj == nullptr || !obj->GetType()->Equals(CSize::typeid))
			return false;

		CSize sz = (CSize)obj;
		return m_width == sz.m_width && m_height == sz.m_height;
	}

	int		CSize::GetHashCode(void) {
		return m_width ^ m_height;
	}

	String ^ CSize::ToString(void) {
		return "{Width=" + m_width.ToString(CultureInfo::CurrentCulture) 
			+ ", Height=" + m_height.ToString(CultureInfo::CurrentCulture) + "}";
	}

#pragma	endregion

}}

File: ShellLib.Entity/Drawing/Size.cs
using System.Diagnostics;
using System.Globalization;

namespace ShellLib.Drawing
{
	[DebuggerDisplay("{Width: {Width}, Height: {Height}}")]
	public struct CSize
	{
		public static readonly CSize Empty = new CSize(0, 0);

		private short _w;
		private short _h;

		public CSize(CCoord c)
		{
			_w = c.X;
			_h = c.Y;
		}

		public CSize(short w, short h)
		{
			_w = w;
			_h = h;
		}

		public short Width { get { return _w; } set { _w = value; } }

		public short Height { get { return _h; } set { _h = value; } }

		public bool IsEmpty { get { return _w == 0 && _h == 0; } }

		public static CSize Add(CSize sz1, CSize sz2)
		{
			return new CSize((short)(sz1.Width + sz2.Width), (short)(sz1.Height + sz2.Height));
		}

		public override bool Equals(object obj)
		{
			if (!(obj is CSize))
				return false;

			var sz = (CSize)obj;

			return _w == sz._w && _h == sz._h;

		}

		public override int GetHashCode()
		{
			return _w ^ _h;
		}

		public static CSize Subtract(CSize sz1, CSize sz2)
		{
			return new CSize((short)(sz1.Width - sz2.Width), (short)(sz1.Height - sz2.Height));
		}

		public override string ToString()
		{
			return "{Width=" + Width.ToString(CultureInfo.CurrentCulture)
				   + ", Height=" + Height.ToString(CultureInfo.CurrentCulture) + "}";
		}

		public static explicit operator CCoord(CSize sz)
		{
			return new CCoord(sz._w, sz._h);
		}

		public static CSize operator +(CSize left, CSize right)
		{
			return Add(left, right);
		}

		public static bool operator ==(CSize left, CSize right)
		{
			return left._w == right._w && left._h == right._h;
		}

		public static bool operator !=(CSize left, CSize right)
		{
			return !(left == right);
		}

		public static CSize operator -(CSize left, CSize right)
		{
			return Subtract(left, right);
		}
	}
}